desktop-widgets/command_divelist.cpp

```diff
@@ -55,6 +89,22 @@ static std::vector<DiveToAdd> removeDives(std::vector<dive *> &divesToDelete)
 		res.push_back(removeDive(d));
 	divesToDelete.clear();
 
+	// We send one dives-deleted signal per trip (see comments in DiveListNotifier.h).
+	// Therefore, collect all dives in a array and sort by trip.
+	std::vector<std::pair<dive_trip *, dive *>> dives;
+	dives.reserve(res.size());
+	for (const DiveToAdd &entry: res)
+		dives.push_back({ entry.trip, entry.dive.get() });
+
+	// Send signals.
+	processByTrip(dives, [&](dive_trip *trip, const QVector<dive *> &divesInTrip) {
```

Comment:
sorry - syntax question... what does the `&` in the `[ ]` mean?
Googling tells me "captures all variables used in the lambda by reference" - that's great. What does that mean again?
Sorry :-(

Reply:
The advantage of lambdas over function pointers is that they can have state. For this purpose, you feed it variables from the containing scope. These variables (=captures) are given in square brackets.

So, for example
```
auto fun = [when] function(dive *d) { return dive->when < when; }
```
would be a function that takes a `dive *` as an input parameter and returns true whenever the start-time is before the local variable `when`:
```
	for_each_dive(i, dive) {
		if (fun(dive))
			return dive;
	}
```
 Obviously, it has to store the `when` somewhere, either by reference or by value. In the case of complex types, the former is more efficient [you don't have to copy the value]. But on the other hand, if the lambda has a longer lifetime, you might end up with dangling references, or changing values.

`[&]` and `[=]` are the lazy options: Don't specify all captured variables, simply say the all used variables are capture by reference (`[&]`) or by value (`[=]`). Hope that clears it up (somewhat).